core/org.wso2.carbon.core/src/main/java/org/wso2/carbon/core/keystore/persistence/KeyStoreRegistryPersistenceManager.java

```diff
@@ -0,0 +1,421 @@
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.core.RegistryResources;
+import org.wso2.carbon.core.internal.CarbonCoreDataHolder;
+import org.wso2.carbon.core.security.KeyStoreMetadata;
+import org.wso2.carbon.core.util.CryptoException;
+import org.wso2.carbon.core.util.CryptoUtil;
+import org.wso2.carbon.core.util.KeyStoreUtil;
+import org.wso2.carbon.registry.api.Association;
+import org.wso2.carbon.registry.api.Registry;
+import org.wso2.carbon.registry.api.RegistryException;
+import org.wso2.carbon.registry.api.Resource;
+import org.wso2.carbon.registry.core.Collection;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
+import org.wso2.carbon.utils.security.KeystoreUtils;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+/**
+ * This implementation handles the keystore storage/persistence related logics in the Registry.
+ */
+public class KeyStoreRegistryPersistenceManager implements KeyStorePersistenceManager {
+
+    private static final Log LOG = LogFactory.getLog(KeyStoreRegistryPersistenceManager.class);
+    private static final String REGISTRY_PATH_SEPARATOR = "/";
+    private static final String ASSOCIATION_TENANT_KS_PUB_KEY = "assoc.tenant.ks.pub.key";
+    private static final String PROP_TENANT_PUB_KEY_FILE_NAME_APPENDER = "tenant.pub.key.file.name.appender";
+
+    /**
+     * Add the key store to the registry.
+     *
+     * @param keyStoreName       Name of the key store.
+     * @param keystoreContent    Content of the key store.
+     * @param provider           Provider of the key store.
+     * @param keyStorePassword   Key Store Password.
+     * @param privateKeyPassword Private key password.
+     * @param tenantId           Tenant ID.
+     * @throws SecurityException If an error occurs while adding the key store.
+     */
+    public void addKeystore(String keyStoreName, byte[] keystoreContent, String provider, String type,
+                            String keyStorePassword, String privateKeyPassword, int tenantId)
+            throws SecurityException {
+
+        Registry registry = getGovernanceRegistry(tenantId);
+        if (isKeyStoreExistsInRegistry(keyStoreName, registry)) {
+            throw new SecurityException("Key store " + keyStoreName + " already available");
+        }
+
+        String path = RegistryResources.SecurityManagement.KEY_STORES + REGISTRY_PATH_SEPARATOR + keyStoreName;
```

Comment:
Lets extract this to a method, since this logic is duplicated across many methods?

Also we can move the initialization of the `path` variable to the closer to its actual use.. ie just before the `registry.put(path, resource)`.. 

Also, if we are `path` used only in a one place, we can avoid declaring the variable and directly use the value where it needed.

Reply:
Addressed with https://github.com/wso2/carbon-kernel/compare/6c4a7e5fc5efca69a114e5d3a08289b8dc809be7..3a9a8a70b486cbcffcd4d4da8c10588bd93f0358